VEP-150: Add cloud-init NoCloud vendor-data support (#151)
kubevirt/enhancements: veps/sig-compute/0150-cloud-init-vendor-data/cloud-init-vendor-data.md

4. **As a multi-tenant platform operator**, I want to inject tenant-specific configuration via vendor-data secrets so that different tenants can have different baseline configurations.

## Repos

- kubevirt/kubevirt

## Design

The implementation adds three new optional fields to both `CloudInitNoCloudSource` and `CloudInitConfigDriveSource` structs in the KubeVirt API:

### CloudInitNoCloudSource

```go
type CloudInitNoCloudSource struct {
    ...
    // New fields for vendor-data
    VendorDataSecretRef  *v1.LocalObjectReference `json:"vendorDataSecretRef,omitempty"`
    VendorDataBase64     string                   `json:"vendorDataBase64,omitempty"`
    VendorData           string                   `json:"vendorData,omitempty"`
}
```

### CloudInitConfigDriveSource

```go
type CloudInitConfigDriveSource struct {
    ...
    // New fields for vendor-data
    VendorDataSecretRef  *v1.LocalObjectReference `json:"vendorDataSecretRef,omitempty"`
    VendorDataBase64     string                   `json:"vendorDataBase64,omitempty"`
    VendorData           string                   `json:"vendorData,omitempty"`
}
```

### Operational Mechanism

> Since `vendorData` is a field on the VM/VMI spec, the standard way for a platform operator to inject it cluster-wide — independently from the VM owner — is through a **MutatingAdmissionPolicy** (or MutatingAdmissionWebhook). The admission controller intercepts VM/VMI creation requests and injects the `vendorData` field before the object is persisted. From the VM owner's perspective, cloud-init simply applies additional platform configuration at boot; they do not need to include vendor-data in their spec at all.

Reviewer: Using `MutatingAdmissionPolicy` is reasonable and will probably solve the operational challenge to decouple cluster level vs VM specifics.

However, if one uses `MutatingAdmissionPolicy` then the existing cloud-init options can be changed anyway without the need of the cloud-init new section.

Also, it is not really a Kubevirt full solution, as the user will need to deploy this admission.

# VEP #150: Add cloud-init NoCloud vendor-data support

## Release Signoff Checklist

Items marked with (R) are required *prior to targeting to a milestone / release*.

- [x] (R) Enhancement issue created, which links to VEP dir in [kubevirt/enhancements](https://github.com/kubevirt/enhancements/issues/150)
- [ ] (R) Target version is explicitly mentioned and approved
- [ ] (R) Graduation criteria filled

## Overview

This VEP proposes adding vendor-data support to the cloud-init NoCloud datasource in KubeVirt. Vendor-data is a standard cloud-init feature that allows cloud providers and operators to inject configuration separately from user-data.

## Motivation

The cloud-init NoCloud datasource supports four types of data:
- **meta-data**: Instance metadata (already supported)
- **user-data**: User-provided configuration (already supported)
- **network-data**: Network configuration (already supported)
- **vendor-data**: Vendor/operator-provided configuration (**not yet supported**)

> Currently, Currently, KubeVirt exposes user-data, network-data, and auto-generates meta-data. Only vendor-data remains unsupported, which limits the ability of cloud providers and platform operators to inject their own configuration without modifying user-data.

Reviewer 2: ```suggestion
KubeVirt currently supports user-data and network-data and auto-generates meta-data. However, vendor-data is not supported, limiting cloud providers and platform operators from injecting custom configuration without modifying user-data.
```

```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  name: my-vm
spec:
  template:
    spec:
      volumes:
        - name: cloudinit
          cloudInitNoCloud:
            userData: |
              #cloud-config
              hostname: my-vm
            networkData: |
              version: 2
              ethernets:
                eth0:
                  dhcp4: true
            vendorData: |
              #cloud-config
              packages:
                - qemu-guest-agent
```

## Alternatives

### Alternative 1: Use user-data for everything

Users could include vendor/platform configuration directly in their user-data.

**Rejected because:**
- Couples user configuration to platform requirements
- Users must know and maintain platform-specific configuration
- Changes to platform defaults require updating all user-data
- Goes against cloud-init's separation of concerns design

### Alternative 2: Use ConfigMaps/Secrets mounted as disks

Platform configuration could be mounted as additional disks that cloud-init reads.

**Rejected because:**
- More complex setup for users
- Doesn't integrate with cloud-init's native vendor-data merging

> - Requires users to configure cloud-init to look for additional datasources

Reviewer 2: In both the selected design and this alternative, users still need to update the VM manifest to include vendor-data—either via cloud-init or by mounting a Secret as a disk. Is the additional complexity here really significant enough to justify rejecting this approach?

No scalability concerns. The vendor-data is processed identically to user-data, which is already proven at scale in production KubeVirt deployments. The data is:
- Stored in etcd as part of the VM spec (same as user-data)
- Mounted in virt-launcher pod (same as user-data secrets)
- Written to ISO at VM start time (same as user-data)

## Update/Rollback Compatibility

### Backward Compatibility

This change is fully backward compatible:
- All new fields are optional
- Existing VMs without vendor-data continue to work unchanged
- No changes to existing API field semantics

### Rollback Safety

If KubeVirt is rolled back to a version without vendor-data support:
- VMs with vendor-data specified will have those fields ignored
- The vendor-data file will not be included in the ISO
- VMs will still boot and function (just without vendor-data)
- No data loss or VM failures

### Upgrade Path

No special upgrade considerations. The feature is additive and optional.

## Functional Testing Approach

### Unit Tests

1. **Reading vendor-data sources**:
   - Test reading inline vendor-data string
   - Test decoding base64 vendor-data
   - Test reading vendor-data from secret reference

>    - Test error handling for invalid base64

Reviewer 2: It's unclear if there is no validation for the vendor-data content, where would the error be surfaced?

KubeVirt currently supports user-data and network-data and auto-generates meta-data. However, vendor-data is not supported, limiting cloud providers and platform operators from injecting custom configuration without modifying user-data.

Cloud-init is designed to merge vendor-data with user-data, allowing operators to provide defaults that users can override. Without vendor-data support, operators must either:
1. Ask users to include platform-specific configuration in their user-data
2. Use post-boot configuration mechanisms that bypass cloud-init

Both alternatives are suboptimal and go against cloud-init's design philosophy.

## Goals

- Add vendor-data support to `CloudInitNoCloudSource` API
- Support three input methods (matching existing patterns for user-data and network-data):
  - Inline string (`vendorData`)
  - Base64 encoded (`vendorDataBase64`)
  - Kubernetes Secret reference (`vendorDataSecretRef`)
- Include vendor-data file in the generated NoCloud ISO
- Implement resource limits (maximum entry counts and value lengths) to prevent abuse
- Maintain full backward compatibility

## Non Goals

> - Adding vendor-data support to ConfigDrive datasource. ConfigDrive support can be added in a separate VEP if there is demand. The NoCloud datasource is more commonly used in KubeVirt environments, and keeping this VEP focused on NoCloud allows for a simpler, more targeted implementation.

Author: Yes, I can add ConfigDrive support! The implementation would be similar - adding the same three fields (vendorData, vendorDataBase64, vendorDataSecretRef) to CloudInitConfigDriveSource. I'll update the VEP and code PR to include ConfigDrive support

- Merging vendor-data with user-data at the KubeVirt level (cloud-init handles this internally)
- Validating vendor-data content format
- Supporting vendor-data for other cloud-init datasources

## Definition of Users

- **Platform operators**: Organizations running KubeVirt clusters who need to inject baseline configuration across all VMs (monitoring agents, security policies, SSH keys, etc.). These operators control the cluster infrastructure and use vendor-data to provide platform-level defaults.

> - **Cloud providers**: Providers offering KubeVirt-based virtualization services who need to provide cloud-specific defaults. Similar to platform operators, they control the service infrastructure and use vendor-data to integrate VMs with platform services.

Reviewer 3: Yeah, sounds reasonable.

   - Add `VendorData` field to internal `CloudInitData` struct
   - Read vendor-data from inline, base64, or secret sources
   - Write `vendor-data` file to the NoCloud ISO

3. **Secret mounting**: Update `pkg/virt-controller/services/rendervolumes.go` to mount `VendorDataSecretRef` as a volume in the virt-launcher pod

4. **ISO generation**: When generating the NoCloud ISO, include `vendor-data` file alongside existing `user-data`, `meta-data`, and `network-config` files

### Entry Limits and Validation

Following the design pattern established in VEP #100 (custom metadata support), this VEP implements resource limits to protect cluster resources and prevent abuse:

- **Content size limit**: Maximum total size of vendor-data content is **256 KB** (262,144 bytes)
- **Enforcement**: Limits are enforced during API validation, preventing VMI creation if vendor-data exceeds the threshold
- **Limit scope**: 
  - Applies to the raw content size of vendor-data (inline string, base64-decoded content, or secret value)
  - Measured after base64 decoding but before ISO generation
  - Consistent with user-data size limits already enforced in KubeVirt

> **Rationale**: Unlike VEP #100's metadata fields which use a key-value map structure (hence the 16-key limit), vendor-data is free-form YAML/text content similar to user-data. Therefore, a content size limit (matching existing user-data limits) is more appropriate than key-count limits. This provides resource protection while allowing operators flexibility to provide comprehensive platform configuration.

Reviewer 3: Sounds reasonable!

# VEP #150: Add cloud-init vendor-data support

## VEP Status Metadata

### Target releases

- This VEP targets alpha for version:
- This VEP targets beta for version:

> - This VEP targets GA for version: v1.9

Reviewer 4: /hold
We need to start with Alpha unless very good reasoning is given

### Release Signoff Checklist

Items marked with (R) are required *prior to targeting to a milestone / release*.

- [x] (R) Enhancement issue created, which links to VEP dir in [kubevirt/enhancements](https://github.com/kubevirt/enhancements/issues/150)
- [ ] (R) Alpha target version is explicitly mentioned and approved
- [ ] (R) Beta target version is explicitly mentioned and approved
- [x] (R) GA target version is explicitly mentioned and approved

## Overview

This VEP proposes adding vendor-data support to the cloud-init NoCloud and ConfigDrive datasources in KubeVirt. Vendor-data is a standard cloud-init feature that allows cloud providers and operators to inject configuration separately from user-data.

## Motivation

> The cloud-init NoCloud and ConfigDrive datasources support four types of data:

Reviewer: You mention four but listed 3.

Author: Thank you for catching that! I apologize for the oversight.... I have corrected it to "three" to match the three items listed below.

- **End users**: VM owners who create and manage VirtualMachine resources. End users typically do not directly interact with vendor-data; instead, they benefit from platform-level configuration provided by operators through vendor-data while focusing their user-data on application-specific needs. In multi-tenant scenarios, namespace-level secrets can allow controlled vendor-data injection.

The typical relationship is: platform operators/providers define vendor-data that applies to VMs, while end users define user-data for their applications. Cloud-init merges these configurations with user-data taking precedence for conflicts.

## User Stories

1. **As a platform operator**, I want a vendor-data field available in the VMI spec so that I can use a MutatingAdmissionPolicy to automatically inject organization-wide defaults (e.g., monitoring agents, security settings) into all VMs without requiring individual VM owners to configure it themselves.

2. **As a cloud provider**, I want to provide cloud-specific networking configuration or monitoring agents via vendor-data so that VMs are automatically integrated with my platform services.

> 3. **As a user**, I want my user-data to remain clean and focused on my application configuration while vendor-data handles platform-specific concerns that I don't need to manage.

Reviewer: I think this one captures the decoupling need. However, I find the Design section not answering how it is resolved from an operational perspective.

What I mean is, if the information is on the VM/VMI spec, it is controlled by the "user", i.e. the namespace admin that creates the VM. How exactly the "provider" will set the relevant content independently from the user?

IIUC, with classic cloud-init, the end user has no direct access to the vendor portion, allowing it to be managed by a higher level authority.

Without this answered, I find the story weak.

Author: Thank you for pointing this out....

The intended mechanism is a MutatingAdmissionPolicy (or MutatingAdmissionWebhook) at the cluster level, where the platform operator intercepts VM/VMI creation requests and injects the `vendorData` field before the object is persisted entirely transparent to the VM owner.

I have added an "Operational Mechanism" subsection to the Design section to make this explicit. I apologize for not having included this from the start.

4. **As a multi-tenant platform operator**, I want to inject tenant-specific configuration via vendor-data secrets so that different tenants can have different baseline configurations.

## Repos

- kubevirt/kubevirt

## Design

The implementation adds three new optional fields to both `CloudInitNoCloudSource` and `CloudInitConfigDriveSource` structs in the KubeVirt API:

### CloudInitNoCloudSource

```go
type CloudInitNoCloudSource struct {
    // Existing fields for user-data
    UserDataSecretRef    *v1.LocalObjectReference `json:"secretRef,omitempty"`
    UserDataBase64       string                   `json:"userDataBase64,omitempty"`
    UserData             string                   `json:"userData,omitempty"`
    
    // Existing fields for network-data
    NetworkDataSecretRef *v1.LocalObjectReference `json:"networkDataSecretRef,omitempty"`
    NetworkDataBase64    string                   `json:"networkDataBase64,omitempty"`

>     NetworkData          string                   `json:"networkData,omitempty"`

Reviewer: I prefer you place instead of these just `...` , these are not part of the proposal.

    
    // New fields for vendor-data
    VendorDataSecretRef  *v1.LocalObjectReference `json:"vendorDataSecretRef,omitempty"`
    VendorDataBase64     string                   `json:"vendorDataBase64,omitempty"`
    VendorData           string                   `json:"vendorData,omitempty"`
}
```

### CloudInitConfigDriveSource

```go
type CloudInitConfigDriveSource struct {
    // Existing fields for user-data
    UserDataSecretRef    *v1.LocalObjectReference `json:"secretRef,omitempty"`
    UserDataBase64       string                   `json:"userDataBase64,omitempty"`
    UserData             string                   `json:"userData,omitempty"`
    
    // Existing fields for network-data
    NetworkDataSecretRef *v1.LocalObjectReference `json:"networkDataSecretRef,omitempty"`
    NetworkDataBase64    string                   `json:"networkDataBase64,omitempty"`

>     NetworkData          string                   `json:"networkData,omitempty"`

Reviewer: I prefer you place instead of these just `...` , these are not part of the proposal.

Author: You are right, I should not have included the existing fields here as they are not part of this proposal. I've replaced them with `...` in both structs. Apologies for the confusion

### Entry Limits and Validation

Following the design pattern established in VEP #100 (custom metadata support), this VEP implements resource limits to protect cluster resources and prevent abuse:

- **Content size limit**: Maximum total size of vendor-data content is **256 KB** (262,144 bytes)
- **Enforcement**: Limits are enforced during API validation, preventing VMI creation if vendor-data exceeds the threshold
- **Limit scope**: 
  - Applies to the raw content size of vendor-data (inline string, base64-decoded content, or secret value)
  - Measured after base64 decoding but before ISO generation
  - Consistent with user-data size limits already enforced in KubeVirt

**Rationale**: Unlike VEP #100's metadata fields which use a key-value map structure (hence the 16-key limit), vendor-data is free-form YAML/text content similar to user-data. Therefore, a content size limit (matching existing user-data limits) is more appropriate than key-count limits. This provides resource protection while allowing operators flexibility to provide comprehensive platform configuration.

> ## API Examples

Reviewer: This section has too many examples IMO.
The new API stands by itself, decoupled from the rest AFAIU. Therefore, I do not see why we need all combinations to be presented (user, network). If you have a real use example, you could have that one posted.

Author: The examples section was unnecessarily verbose. I've trimmed it down to two focused examples  both demonstrating the real-world platform operator use case with `vendorDataSecretRef`, one for NoCloud and one for ConfigDrive. Thank you for the guidance

- **Content size limits**: The 256 KB maximum size for vendor-data prevents resource exhaustion from excessively large payloads. This matches existing user-data size limits in KubeVirt.
- **API validation**: Size limits are enforced during API validation, preventing VMI creation if vendor-data exceeds the threshold.
- **Processing overhead**: Vendor-data is processed identically to user-data, which is already proven at scale in production KubeVirt deployments. The data is:
  - Stored in etcd as part of the VM spec (same as user-data)
  - Mounted in virt-launcher pod (same as user-data secrets)
  - Written to ISO at VM start time (same as user-data)
- **Performance impact**: The 256 KB size limit ensures cloud-init processing remains performant and does not impact VM startup times.
- **Resource protection**: Size limits protect against both accidental misconfiguration and intentional abuse of the vendor-data feature.

## Update/Rollback Compatibility

### Backward Compatibility

This change is fully backward compatible:
- All new fields are optional
- Existing VMs without vendor-data continue to work unchanged
- No changes to existing API field semantics

### Rollback Safety

> Rolling back KubeVirt to a version without vendor-data support is safe and causes graceful degradation:

Reviewer: AFAIK Kubevirt does not support rollback (to a previous version).
I am not even sure why we have this section here.

If we had support for this, a new VM with the new data could have been migrated to an older node, and I am unsure what would have happened. But again, we do not support this flow.

@0xFelix , @vladikr , anyone, WDYT?

Author: You are right..... I was not aware that KubeVirt does not support rollback to a previous version, so this section did not belong here. I have removed it. Thank you for the clarification

- Consider testing rollback scenarios if vendor-data provides critical platform functionality

### Upgrade Path

No special upgrade considerations. The feature is additive and optional.

## Functional Testing Approach

### Unit Tests

1. **Reading vendor-data sources** (NoCloud and ConfigDrive):
   - Test reading inline vendor-data string
   - Test decoding base64 vendor-data
   - Test reading vendor-data from secret reference
   - Test error handling for invalid base64

2. **ISO generation**:
   - Test that vendor-data file is included in NoCloud ISO when provided
   - Test that vendor_data.json file is included in ConfigDrive ISO when provided
   - Test that vendor-data file is omitted when not provided
   - Test combined user-data, network-data, and vendor-data

3. **Secret resolution**:
   - Test resolving vendor-data from mounted secret for NoCloud
   - Test resolving vendor-data from mounted secret for ConfigDrive
   - Test handling missing secret gracefully

### Functional/E2E Tests

1. **VM boot with vendor-data**:
   - Create VM with vendor-data (NoCloud)
   - Create VM with vendor-data (ConfigDrive)
   - Verify VM boots successfully

>    - Verify cloud-init processes vendor-data

Reviewer: I think the next one is covering this one implicitly already.
As we focus on user scenarios, and the cost of these tests are high, I think this one is not needed.

Author: That makes sense..... I had not thought about the implicit coverage. I have removed item 1 as the content verification test already covers it. Thank you for the suggestion.

2. **Vendor-data content verification**:
   - Create VM with known vendor-data content
   - Verify vendor-data file exists in NoCloud/ConfigDrive ISO
   - Verify content matches what was specified

3. **Secret-based vendor-data**:
   - Create secret with vendor-data
   - Create VM referencing the secret (both NoCloud and ConfigDrive)
   - Verify vendor-data is correctly injected

## Implementation History

- 2025-12-03: Initial implementation PR opened: https://github.com/kubevirt/kubevirt/pull/16278
- 2025-12-09: VEP created: https://github.com/kubevirt/enhancements/issues/150
- 2025-12-25: Added ConfigDrive vendor-data support to implementation

## Graduation Requirements

> ### GA

Reviewer: @vladikr has raised a concern about this being pushed directly to GA.

I believe this was done as a follow up to a previous [similar addition](https://github.com/kubevirt/enhancements/pull/101) and from the same reasons.

I can think about one reason this merits going through an Alpha or Beta stage: Check actual adoption.
We should have asked the same on the previous VEP.

In case there are existing downstream projects that use this today, it can be used as a reasoning for direct GA.

Reviewer: Worth mentioning that the [issue tracker](https://github.com/kubevirt/enhancements/issues/150) is tracking this VEP as Alpha.

# VEP #150: Add cloud-init vendor-data support

## VEP Status Metadata

### Target releases

- This VEP targets alpha for version:
- This VEP targets beta for version:
- This VEP targets GA for version: v1.9

### Release Signoff Checklist

Items marked with (R) are required *prior to targeting to a milestone / release*.

- [x] (R) Enhancement issue created, which links to VEP dir in [kubevirt/enhancements](https://github.com/kubevirt/enhancements/issues/150)
- [ ] (R) Alpha target version is explicitly mentioned and approved
- [ ] (R) Beta target version is explicitly mentioned and approved
- [x] (R) GA target version is explicitly mentioned and approved

## Overview

This VEP proposes adding vendor-data support to the cloud-init NoCloud and ConfigDrive datasources in KubeVirt. Vendor-data is a standard cloud-init feature that allows cloud providers and operators to inject configuration separately from user-data.

## Motivation

The cloud-init NoCloud and ConfigDrive datasources support four types of data:

> - **meta-data**: Instance metadata (already supported)

Reviewer 3: Can you leave metadata out since it is not merged yet? (VEP https://github.com/kubevirt/enhancements/issues/100)

Author: Done ....removed the meta data bullet from the list since VEP #100 is not merged yet

3. **Secret resolution**:
   - Test resolving vendor-data from mounted secret for NoCloud
   - Test resolving vendor-data from mounted secret for ConfigDrive
   - Test handling missing secret gracefully

### Functional/E2E Tests

1. **VM boot with vendor-data**:
   - Create VM with vendor-data (NoCloud)
   - Create VM with vendor-data (ConfigDrive)
   - Verify VM boots successfully
   - Verify cloud-init processes vendor-data

2. **Vendor-data content verification**:
   - Create VM with known vendor-data content
   - Verify vendor-data file exists in NoCloud/ConfigDrive ISO
   - Verify content matches what was specified

3. **Secret-based vendor-data**:
   - Create secret with vendor-data
   - Create VM referencing the secret (both NoCloud and ConfigDrive)
   - Verify vendor-data is correctly injected

## Implementation History

- 2025-12-03: Initial implementation PR opened: https://github.com/kubevirt/kubevirt/pull/16278
- 2025-12-09: VEP created: https://github.com/kubevirt/enhancements/issues/150
- 2025-12-25: Added ConfigDrive vendor-data support to implementation

## Graduation Requirements

### GA

**Rationale for direct GA graduation:**

> Following the precedent set by VEP #100 (custom metadata support), this enhancement proposes direct GA graduation without an Alpha/Beta phase. The rationale includes:

Reviewer 3: Needs to be changed a little I'd say, VEP #100 is not in yet.

Author: Done .....removed the VEP #100 precedent reference. The rationale now stands on its own merits without citing an unmerged VEP
